### ISSUE TITLE
Please don't kill process on uncaught exception

### ISSUE THREAD

**Issue by demongloom** [segmentio/nightmare#586]
```
process.on('uncaughtException', function(err) {
    console.error(err.stack);
    endInstance(self);
    // This allows any user defined 'uncaughtException' handlers
    // to run before exiting
    process.nextTick(function() {
        process.exit(1);
    });
  });
```

Please give to nightmare consumer decide when to kill process.
Thanks

**Comment by rosshinkley** [comment]
This came up in #558, and I'll quote @Mr0grog:

> So… it would be a very bad idea not to exit the process at this point. Electron’s underlying IPC machinery is now in a bad state, so further communication between the your node.js process and the “browser” is not reliable.

I'm wondering now is if instead of exiting, the Nightmare instance should be put into an error state where any further calls would cause exceptions as the IPC parts are no longer reliable.

All of that being said: uncaught exceptions are bad.  Why not add handling for the exception?

**Comment by Mr0grog** [comment]
> I'm wondering now is if instead of exiting, the Nightmare instance should be put into an error state where any further calls would cause exceptions as the IPC parts are no longer reliable.

I definitely think this would be great. But some thoughts: there are two different kinds of `uncaughtException` handlers and I think the story is a little different for each.

1. Exceptions in the child (electron) process. This is what #558 was about. I think the reality here is that when this happens, we pretty much have to assume everything in the process (and therefore in the Nightmare instance controlling it) is hosed. I think you’re on the right track here, @rosshinkley:

    1. Treat this like an error in any action, stopping all running queues, and calling the callback/rejecting the promise for each of them with information about the exception.

    2. Marking the instance as dead so any later calls to `run/then/end` either throw or reject (I’d lean towards reject, but throwing might make sense, too). The same goes for when the Electron process crashes, which is currently totally silent. 😱 Arguably, we should be doing this when the instance is explicitly ended, too (IIRC, you can currently call `run/then` again after the process is ended and have unclear problems).

2. Exceptions in the parent (user) process (the handler @demongloom cited). As I’m thinking about this one more critically now, **I’m not sure we should be handling this event at all:**

    - Handling it is not great practice for library code (as opposed to user code) because it changes the behavior of the process. We “fixed” this in #554, but this gets in the user’s way if they want to not crash on an exception (rarely ever a good idea, but *sometimes* there are legitimate reasons). In hindsight, the better fix for #554 might have been to just remove the handler altogether.
    - Handling it in order to clean up the electron process is unnecessary because we already trap all the possible close/exit/signal events and cleanup there. Whether the system exits because of an uncaught exception or the user explicitly does so in their own code, we’re still covered without trapping this event.
    - Presumably the handler was originally added to ease debugging, but what about cases where Electron is a useful additional piece of functionality in a larger app, rather than the central purpose of someone’s app? It’s not great to just be spewing stack traces to the console if they’re trying to redirect errors somewhere else (logs, a remote service, etc). (If we really like this functionality, maybe send it to `debug` and then only exit the process if there are no other uncaught exception handlers?)

    To some extent, I simply did wrong by adding the `endInstance` call to the existing uncaught exception handler in #487. We don’t have to assume, unlike with the child process, that Nightmare’s internals are hosed here. It’s plenty likely that the exception came from code that didn’t have anything to do with Nightmare.

**Comment by rosshinkley** [comment]
Fixed with #589.  Closing.
